### PR TITLE
feat: consolidate HTTP and websockets

### PR DESCRIPTION
**Pull request by jonty-kennedy-trendsales**
* Removed all HTTP calls from lobbies except fetching list of available lobbies to join (which I think makes sense to keep as HTTP) so creating and joining lobby are all handled over websockets
* Consolidated lobby repository and user connections repository
* Implemented a ping pong mechanism to handle removing players who have lost connection (with a delay in case of temporary disruptions)
  * Handled closing lobbies with no players left
* Ran linting and sorted out inconsistent variables (game_id + lobby_id -> game_id, player_name + username -> player_name, etc.)

Demo of new changes:

https://github.com/user-attachments/assets/407dee90-7a64-40cc-b65a-53a6e51f8e45

And init game still works:

https://github.com/user-attachments/assets/b50c1e0f-5178-4287-bc5f-5bfd8b824b75

### CODE REVIEW

File: backend/game_flow/game_flow_websocket.py [hunk 1]
```diff
@@ -35,113 +62,112 @@ def init_game(host_id, game_id):
     game_state.active = True
 
 
-def join_game(player_id, player_name, game_id, websocket):
-    if game_id not in player_connections_by_game_id:
-        message = 'game does not exist'
+def join_game(websocket_message: WebsocketMessage, connection):
```
Comment (jonty-kennedy-trendsales) [inline backend/game_flow/game_flow_websocket.py:65]: I'll refactor all the exception throwing in here later

File: backend/game_flow/game_flow_websocket.py [hunk 2]
```diff
@@ -35,113 +62,112 @@ def init_game(host_id, game_id):
...
+            message=f"Player {player_name} entered the game",
+            playerId=player_id,
+            playerName=player_name,
+            lobby=serialise_lobby(lobby),
+        )
 
-    websocket.send(json.dumps(response.__dict__))
+        for player in lobby["players"].values():
+            if player["connection"] is not None:
+                player["connection"].send(json.dumps(response.__dict__))
```
Comment (jonty-kennedy-trendsales) [inline backend/game_flow/game_flow_websocket.py:99]: Will replace this will broadcast function later

File: backend/utils/ping_client.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+lobby_repository = LobbyRepository()
+
+
+def ping_client(websocket, player_id, game_id):
+    while True:
+        time.sleep(2)
+        try:
+            ping_message = json.dumps({"method": "ping"})
+            websocket.send(ping_message)
+            websocket.ping_time = time.time()
```
Comment (andrevigneaux) [inline backend/utils/ping_client.py:14]: what is this one for?

File: backend/utils/ping_client.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+            time.sleep(2)
+            if not websocket.pong_received:
```
Comment (andrevigneaux) [inline backend/utils/ping_client.py:17]: how will this one be false?

File: frontend/scenes/JoinLobbyScene.js [hunk 5]
```diff
@@ -103,33 +103,51 @@ export default class JoinLobbyScene extends Phaser.Scene {
...
+        game_id: gameId,
+        payload: {
+          player_name: playerName,
+        },
+      };
+
+      ws.send(JSON.stringify(message));
+    };
+
+    ws.onmessage = (event) => {
```
Comment (andrevigneaux) [inline frontend/scenes/JoinLobbyScene.js:126]: maybe I'm missing something, but shouldn't we check that the method is "join"? considering we also have the pong message coming now and then?

File: frontend/scenes/StartScene.js [hunk 6]
```diff
@@ -47,80 +47,80 @@ export default class StartScene extends Phaser.Scene {
...
+    ws.onerror = (error) => {
+      console.error("WebSocket error:", error);
+    };
+
+    ws.onopen = () => {
+      const createMessage = this.buildCreateWSMessage(playerName);
+      ws.send(JSON.stringify(createMessage));
+    };
+
+    ws.onmessage = (event) => {
```
Comment (andrevigneaux) [inline frontend/scenes/StartScene.js:82]: same as the other comment

**Review by andrevigneaux — APPROVED** [review]
looks good! left some questions

File: backend/utils/ping_client.py (hunk 4 above)
Comment (jonty-kennedy-trendsales) [reply backend/utils/ping_client.py:17]: Oops I did some refactoring after I made the video demo and must've of accidentally removed it. Same with the comment above, that one should've been removed instead

Comment (jonty-kennedy-trendsales) [reply backend/utils/ping_client.py:17]: Fixed

File: backend/utils/ping_client.py (hunk 3 above)
Comment (jonty-kennedy-trendsales) [reply backend/utils/ping_client.py]: Removed

File: frontend/scenes/JoinLobbyScene.js (hunk 5 above)
Comment (andrevigneaux) [reply frontend/scenes/JoinLobbyScene.js:126]: aah I see the sequence of events now 👍

**Pull request merged**